amethyst_renderer/src/pipe/pass.rs

```diff
@@ -1,66 +1,137 @@
 //! Types for constructing render passes.
 
 #![allow(missing_docs)]
-
 use std::fmt::{Debug, Formatter, Result as FmtResult};
-use std::sync::Arc;
+use std::marker::PhantomData;
+
+use rayon::iter::ParallelIterator;
+use rayon::iter::internal::UnindexedConsumer;
+use rayon_core;
+use specs::SystemData;
 
 use error::Result;
 use pipe::{Effect, NewEffect, Target};
-use scene::{Model, Scene};
 use types::{Encoder, Factory};
 
-pub trait Pass: Send + Sync {
-    fn compile(&self, effect: NewEffect) -> Result<Effect>;
-    fn apply(&self, enc: &mut Encoder, effect: &mut Effect, scene: &Scene, model: &Model);
+pub struct Supplier<'a> {
+    encoders: *mut [Encoder],
+    effects: *mut [Effect],
+    pd: PhantomData<&'a ()>,
 }
 
-#[derive(Clone)]
-pub(crate) struct Description(Arc<Pass>);
+impl<'a> Supplier<'a> {
+    fn new(encoders: &'a mut [Encoder], effects: &'a mut [Effect]) -> Self {
+        Supplier {
+            encoders: encoders,
+            effects: effects,
+            pd: PhantomData,
+        }
+    }
 
-impl Description {
-    pub fn new<P: Pass + 'static>(pass: P) -> Self {
-        Description(Arc::new(pass))
+    fn index(&self) -> usize {
+        rayon_core::current_thread_index().expect("Should be called from ThreadPool")
     }
 
-    pub fn compile(self, fac: &mut Factory, out: &Target) -> Result<CompiledPass> {
-        let eb = NewEffect::new(fac, out);
-        let effect = self.0.compile(eb)?;
-        Ok(CompiledPass {
-            effect,
-            inner: self.0,
-        })
+    unsafe fn get(&self) -> (&mut Encoder, &mut Effect) {
```

Comment:
I know this isn't part of the public API but just for the sake of future maintainers can we document why this is unsafe and what conditions need to be guaranteed by the developer before they call this?